File: LectureCodes/IterativeMethods/bisect/Matlab/bisect.m
function x = bisect(@{\com{F}}@,a,b,tol)
% Searching zero of \Blue{$F$} in \Blue{$[a,b]$} by bisection
if (a>b), t=a; a=b; b=t; end;
fa = F(a); fb = F(b);
if (fa*fb>0), error('f(a), f(b) same sign'); end;
if (fa > 0), v=-1; else v = 1; end
x = 0.5*(b+a);
while((b-a > tol) && @\com{((a<x) \& (x<b))}@) @\Label[line]{bs:2}@
  if (v*F(x)>0), b=x; else a=x; end;
  x = 0.5*(a+b)
end  

File: LectureCodes/IterativeMethods/bisect/Python/bisect.py
import numpy as np


def bisect(f, a, b, tol):
    """ Searching zero of f in [a, b] by bisection """
    if a > b:
        a, b = b, a
    fa = f(a)
    fb = f(b)
    if fa * fb > 0:
        raise ValueError('f(a) and f(b) have the same sign')
    v = -1 if fa > 0 else 1
    x = 0.5 * (a + b)
    while b - a > tol and a < x < b:
        if v * f(x) > 0:
            b = x
        else:
            a = x
        x = 0.5 * (a + b)
    return x


def example():
    def f(x): return np.sin(x) * np.cos(x)
    print('sin(x) * cos(x) has zero at {}'.format(bisect(f, 1, 2, 1e-6)))


if __name__ == '__main__':
    example()
